### PR TITLE
Python in mantid

### LINKED ISSUE #25796: Rewrite Mantid Courses Against Workbench
Might require that more plotting functionality (SliceViewer etc) is available first. The courses need to be updated and re-written against `MantidWorkbench` - this would be a useful avenue for showing off the workbench and discontinuing `MantidPlot`.

Look at other tickets marked "Workshop" as part of this

### LINKED ISSUE #28509: Convert Python in Mantid course to WB
This is currently broad but will encompass a few things:
* What things can WB currently not do, see [here](https://www.mantidproject.org/Matrix_and_Plot_Control).
* Should the tutorial be moved to the docs from the wiki?
* Moving from Plot to WB

[Current Course](https://www.mantidproject.org/Python_In_Mantid)

### PR DESCRIPTION
**Pull request by DanielMurphy22**
**Description of work.**

Update the Current Python in Mantid course by moving it to the user docs and making it accurate for Workbench

**To test:**
Build docs.
* Check page order, links and formatting.
* Check I have had the courses and solutions checked by other devs!
* Check all images and plot directives appear
* Check the exercises
* Check the plot script generation in Workbench gives the link to the scripting_plots webpage (moved from `docs/source/plotting/scripting_plots` to `docs/source/tutorials/python_in_mantid/plotting/02_scripting_plots`)

Is this fixed? https://forum.mantidproject.org/t/integration-algorithm-not-giving-sensible-or-reproducible-results/666/6
Fixes #28509
Fixes #25796 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/tutorials/python_in_mantid/further_alg_ws/05_settings.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+All settings specified in these files are accessible via python through the config object, which can be imported from the mantid module. For example, to access the above setting do
+
+.. code-block:: python
+
+	from mantid import config
+	 
+	default_inst_name = config['default.instrument']
+
+It looks like and can be used just like a Python dictionary where the key on the left-hand side of the equals in the properties file is used to access the value in the "dictionary".
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/05_settings.rst:31]:
```suggestion
The config object behaves like a Python dictionary, where each setting in the properties file is used as a key to obtain its value from the config object. 
```

File: docs/source/tutorials/python_in_mantid/further_alg_ws/05_settings.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+The dictionary-like manner can also be used to update settings, i.e.
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/05_settings.rst:34]:
```suggestion
The dictionary-like manner can also be used to update settings, e.g.,
```

File: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+	ws = Load(Filename="HRP39182.RAW")
+	for i in range(ws.getNumberHistograms()):
+	    y = ws.readY(i)
+	    x = ws.readX(i)
+	    e = ws.readE(i)
+
+Warning
+-------
+**Be careful**: the outputs of *read* (y,x,e) are only **views into the data held by the workspace,** `ws`. If `ws` is deleted, the contents of x,y,e will be nonsense (the random contents of the memory locations formerly used for `ws`).
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst:37]:
```suggestion
**Be careful**: the outputs of *read* (y,x,e) are only **views into the data held by the workspace,** `ws`. If `ws` is deleted, the contents of x,y,e will be invalid (the random contents of the memory locations formerly used for `ws`).
```

File: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+If you need x,y,e to persist longer than ws, use the *extract*, which creates a copy of the data in `ws` into y,x,e.
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst:38]:
```suggestion
If you need x,y,e data to persist longer than the workspace use the *extract* methods, which create a copy of the data in `ws`.
```

File: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+The data from all spectra can be obtained as a mutable multi-dimensional array in one-call using the extract methods.
+
+.. code-block:: python
+
+	ws = Load(Filename="HRP39182.RAW")
+	x = ws.extractX()
+
+	print(x.shape)
+	print(y.shape)
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst:51]: Are these print statements correct, as we don't have a `y` or `e` variable, or maybe I'm missing something.

File: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    ws.extractX()[5, :]
+
+    xmat = ws.extractX(); x = xmat[5, :]
+
+
+Nested Looping
+==============
+
+This allows access to the individual bins in each spectrum. e.g. To sum the y-values in each spectrum:
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst:68]:
```suggestion
This allows access to the individual bins in each spectrum. E.g. to sum the y-values in each spectrum:
```

File: docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst [hunk 7]
```diff
@@ -0,0 +1,75 @@
+.. _03_instrument_view:
+
+=======================
+Instrument View Control
+=======================
+
+#The MantidPlot instrument window viewing the GEM instrument
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst:7]: Is this supposed to be here?

File: docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+* When a workspace is loaded for a particular instrument, a virtual geometry is also created. The instrument view window within MantidPlot uses 3D rendering to visualize this geometry.
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst:9]: How come this focouses on MantidPlot?

File: docs/source/tutorials/python_in_mantid/plotting/04_sliceviewer.rst [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    #Plotting a line integration
+    lv = svw.showLine(start=[-7, 6], end=[7, 6], width=0.1)
+
+    # Overplotting a peaks workspace
+    peaksws = Load('peaks_qLab.nxs')
+    sv = svw.getSlicer()
+    sv.setPeaksWorkspaces(['peaksws'])
+
+For more details on the available functionality see the SliceViewer documentation
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/plotting/04_sliceviewer.rst:33]: Might be nice to include a link to the documentation.

File: docs/source/tutorials/python_in_mantid/script_generation/02_extracting_alg_prop.rst [hunk 10]
```diff
@@ -0,0 +1,28 @@
+.. _02_extracting_alg_prop:
+
+===============================
+Extracting Algorithm Properties
+===============================
+
+After an algorithm has been executed the properties remain available to access as long as the produced workspace is in existence.
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/script_generation/02_extracting_alg_prop.rst:7]:
```suggestion
After an algorithm has been executed, the properties remain accessible for as long as the output workspace exists.
```

File: docs/source/tutorials/python_in_mantid/script_generation/04_pim_ex_2.rst [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+
+
+A - Processing ISIS Data
+========================
+
+Running an analysis Manually
+----------------------------
+
+Here you are going to perform a number of steps in the Mantid user interface to correct some LOQ data by the incident beam monitor. For more information in a specific algorithm, click on its name.
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/script_generation/04_pim_ex_2.rst:24]:
```suggestion
Here you are going to perform a number of steps in the Mantid user interface to correct a LOQ dataset using the incident beam monitor. For more information in a specific algorithm, click on its name.
```

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Looks good, just some minor rewording in places. Will test the doc pages and links now

File: docs/source/tutorials/python_in_mantid/script_generation/02_extracting_alg_prop.rst [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+To access input properties, use the :ref:`workspace history <Workspace-Workspace_History>`. Using this handle allows the properties to be extracted by their name using the `getPropertyValue()` function:
+
+.. code-block:: python
+
+    sample = Load('MAR11015.raw')
+    algorithm = sample.getHistory().lastAlgorithm()
+    filename = algorithm.getPropertyValue("Filename")
+
+For some algorithms, such as :ref:`algm-Rebin`, you can get properties without accessing algorithm history.
+.. code-block:: python
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/script_generation/02_extracting_alg_prop.rst:25]: For me this code block isn't formatting correctly. May need an extra space above

File: docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    print(instrument.getName())
+
+:ref:`SpectrumInfo`, :ref:`DetectorInfo` and :ref:`ComponentInfo` have many other features:
+
+.. code-block:: python
+
+    info = ws.spectrumInfo()
+    print(info.hasDetectors(0))
+
+Other :ref:`Workspace Properties <Workspace-Properties_Workspaces>`
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst:77]: I would probably removes these lines of getMemorySize and getHistory and getRun, they don't seem relevant to the discussion and they are stated but not really explained. It looks a bit disjointed in the docs.

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Documentation pages looks fine for the most part, just two things I saw that looked a bit off. I'll check the exercises on Monday.

File: docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+=================
+
+There is a lot of information about workspaces that can be accessed.
+
+Below is a brief overview, but you may prefer:
+
+* :ref:`An exhaustive list <mantid.api.MatrixWorkspace>`
+* :ref:`A wider explanation <MatrixWorkspace>`
+
+getSpectrum for info on the structure of a workspace eg. the spectrum number related to workspace index 0:
```
[the remark this answers is not in the document]
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst:51]: I couldn't find a suitable page to link it to...

File: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst (hunk 5 above)
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst:51]: I'd copied that over from here and for some reason left out the extractY and extractE lines!

File: docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst (hunk 7 above)
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/plotting/03_instrument_view.rst:7]: So the instrument and sliceviewer aren't accessible from a script in workbench yet, so I copied over these pages from the old PythonInmantid course, but I haven't edited them at all! I've left them hidden from the doctree, but I guess I could be convinced to remove them. I can definitely remove any mention of MantidPlot.

File: docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst (hunk 13 above)
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/further_alg_ws/03_matrix_ws_attributes.rst:77]: Good idea. I was a little unsure here, but wanted to give lots of resources. But I agree this was probably an overload!

**Comment by DanielMurphy22** [comment]
@StephenSmith25 All your comments have been addressed in the latest commit. I've left 2 unresolved so you can see my comments. I haven't removed the hidden Sliceviewer and InstrumentViewer pages yet, but I can do if you think I should.

File: docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    fig, axes = plt.subplots(edgecolor='#ffffff', num='run_lambda_summed-1', subplot_kw={'projection': 'mantid'})
+    axes.plot(run_lambda_summed, color='#1f77b4', label='run_lambda_summed: spec 1', linewidth=1.0, specNum=1)
+    axes.plot(run_monitors_lambda_rebinned, color='#ff7f0e', label='run_monitors_lambda_rebinned: spec 1', linewidth=1.0, specNum=1)
+    axes.plot(normalized, color='#2ca02c', distribution=False, label='normalized: spec 1', linewidth=1.0, specNum=1)
+    axes.set_title('run_lambda_summed')
+    axes.set_xlabel('Wavelength ($\AA$)')
+    axes.set_ylabel('($\AA$)$^{-1}$')
+    axes.set_xlim([2.405, 4.5])
+    axes.set_yscale('log')
+    axes.legend() #.set_draggable(True)
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst:132]: I feel when people are reading this solution they'll possibly wonder why this comment is here. But I have no issues with leaving it in if they are supposed to use it.
```suggestion
    axes.legend()
```

File: docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+    Rebin(InputWorkspace='Small_Angle', OutputWorkspace='Small_Angle', Params= params)
+
+    # Extract the Spectrum for correcting the data
+    ExtractSingleSpectrum(InputWorkspace='Small_Angle_monitors', OutputWorkspace='Small_Angle_monitors', WorkspaceIndex=1)
+
+    # Correct the data by dividing by the monitor spectrum
+    Divide(LHSWorkspace='Small_Angle', RHSWorkspace='Small_Angle_monitors', OutputWorkspace='Corrected_data')
+
+
+B - Plotting ILL Data
```
Comment (StephenSmith25) [inline docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst:46]: There’s two code blocks in this solution and only the python bit gets created in the html page, which starts with ADS.retrieve('164198), however at this point that workspace doesn't exist. So maybe the first call needs to be load.

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Exercises work, just two minor thing and its good to go.

File: docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst (hunk 16 above)
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst:46]: I understand that this "doesn't work" as it doesn't actually load the data, but exercise B was simply about generating a script to produce a plot. This will work for the people taking the course as they will already have the data loaded in the Workspaces Toolbox. I also want to display the plot so in the plot directive below this code-block I actually do load the data.

File: docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst (hunk 15 above)
Comment (DanielMurphy22) [reply docs/source/tutorials/python_in_mantid/solutions/02_pim_sol.rst:132]: Again this is because the solution comes from a generated script. But `.set_draggable()` is from a newer mpl version than our build servers, so I had to comment it out. I left the comment here so that the solution matches the script generation that the user will produce.

**Review by StephenSmith25 — APPROVED** [review]
Good to go 👍
